src/express.re

```diff
@@ -1,43 +1,149 @@
+  external json : t => Js.Json.t => done_ = "" 
+    [@@bs.send] [@@ocaml.deprecated "Use sendJson instead`"];
 };
 
-module Express = {
+module Next : {
+  type content;
+
+  type t = (Js.undefined content => done_) [@bs];
+
+  let router : Js.undefined content;  
+  /** value to use as [next] callback argument to skip middleware 
+      processing. */ 
+
+  let error : Error.t => Js.undefined content;
+  /** [error e] returns the argument for [next] callback to be propagate 
+      error [e] through the chain of middleware. */
+
+} = {
+
+  type content;
+
+  type t = (Js.undefined content => done_) [@bs];
+
+  external castToContent : 'a => content = "%identity";
+
+  let router = 
+    Js.Undefined.return (castToContent "router");
+
+  let error (e:Error.t) =>
+    Js.Undefined.return (castToContent e);
+
+};
+
+module Middleware = {
+  type next = Next.t;
+  type t; 
+      /* Middleware abstract type which unified the various way an
+         Express middleware can be constructed:
+         {ul
+         {- from a {b function} using the [ofFunction]}
+         {- from a Router class}
+         {- from an App class}
+         {- from the third party middleware modules}
+         }
+         For each of the class which implements the middleware interface
+         is JavaScript, one can use the "%identity" function to cast
+         it to this type [t] */
+  type f = Request.t => Response.t => next => done_;
+  
+  external fromF: f => t = "%identity"
+    [@@ocaml.doc " Create a Middleware from a function " ]; 
+  
+  type errorF = Error.t => Request.t => Response.t => next => done_;
+  
+  external fromErrorF: errorF => t = "%identity"
+    [@@ocaml.doc " Create a Middleware from an error function " ]; 
+};
+
+/* Generate the common Middleware binding function for a given
+ * type. This Functor is used for the Router and App classes. */
+module MakeBindFunctions (T: {type t;}) => {
+  external use : T.t => Middleware.t => unit = "" [@@bs.send];
+  external useOnPath : T.t => path::string => Middleware.t => unit = "use" [@@bs.send];
+  external get : T.t => path::string => Middleware.t => unit = "" [@@bs.send];
+  external getN : T.t => path::string => array Middleware.t => unit = "get" [@@bs.send];
+};
+
+module App = {
   type t;
-  type middlewareT = Request.t => Response.t => Next.t => unit;
-  external express : unit => t = "" [@@bs.module];
-  external use : t => middlewareT => unit = "" [@@bs.send];
-  external static : path::string => middlewareT = "" [@@bs.module "express"];
-  external get : t => string => ('a => Response.t => unit) => unit = "" [@@bs.send];
+
+  include MakeBindFunctions { type nonrec t = t; };
+  
+  external make : unit => t = "express" [@@bs.module]
+    [@@ocaml.doc " [make ()] create an App class "]; 
+  
+
+  external toMiddleware : t => Middleware.t = "%identity"
+    [@@ocaml.doc " [toMiddleware app] casts an App  to a Middleware type "]; 
```

Comment:
It might not matter much in this particular case, but in general I think it is important to make the distinction and have the option when it does matter. Like if you have different memory representations for the same basic data structure with very different performance profiles, where converting between them involves a very expensive copy. Then you don't want people to think it's cheap and easy to convert back and forth between them to access operations that are just a bit more convenient for one representation or the other.

But as I said, there's not necessarily anything wrong with abstracting that away in specific cases. I think `as*` is a more specific `to*` with an additional perf constraint, so it should be abstracted towards `to*`, and we really should avoid just throwing `as`es around (oh the puns I could make, but I'm a grown-up so I've decided not to).